### PR TITLE
[ndslice] creation/allocation utilities

### PR DESCRIPTION
**Pull request by 9il**
Allocation utilities was not implemented before because I needed a time to get real world experience with ndslice and allocators.

In addition, this PR fixes ugly unittests like 
```D
auto slice = (3*4*5*6).iota.array.sliced(3, 4, 5, 6);
```
Now they looks like 

```D
auto slice = iotaSlice(3, 4, 5, 6, 7).slice;
```

In addition, `iotaSlice` is faster than sliced `iota`, it has zero size and zero cost.

### Mir
https://github.com/DlangScience/mir/pull/38
https://github.com/DlangScience/mir/pull/42
https://github.com/DlangScience/mir/pull/44

#### `slice`

- `SliceException`
- `shape`
- `slice` (`createSlice` in mir PR)
- `makeSlice`
- `ndarray`
- `makeNdarray`

#### `selection`

- `iotaSlice`/`IotaSlice`

### CODE REVIEW

**Comment by 9il** [comment]
ping @John-Colvin @wilzbach @JackStouffer

File: std/experimental/ndslice/slice.d [hunk 1]
```diff
@@ -581,6 +492,314 @@ pure nothrow unittest
...
+    }
+    else
+    {
+        import std.algorithm.iteration: map;
+        return makeArray!(typeof(makeNdarray!T(alloc, slice.front)))(alloc, slice.map!(a => makeNdarray!T(alloc, a)));
+    }
+}
+
+///
+unittest
```
Comment (wilzbach) [inline std/experimental/ndslice/slice.d]: needs `@nogc`

Comment (9il) [inline std/experimental/ndslice/slice.d]: thanks!

**Comment by wilzbach** [comment]
> ping @John-Colvin @wilzbach @JackStouffer

I already tried my best to review the changes at `mir` - and I am still happy with the changes :)
Btw @9il I still think that diffing the changes between `mir` and `std.experimental.ndslice` will require quite some work in the future, especially when more changes come in at mir.

**Comment by 9il** [comment]
> Btw @9il I still think that diffing the changes between mir and std.experimental.ndslice will require quite some work in the future, especially when more changes come in at mir.

I am thinking about automatic convetion tool ;)

**Comment by 9il** [comment]
@wilzbach what do you think about to rename `createSlice` to `slice`? We already have for example `array` and `makeArray`, `ndarray` and `makeNdarray`.

File: std/experimental/ndslice/iteration.d [hunk 2]
```diff
@@ -1083,8 +1075,8 @@ body
 @safe @nogc pure nothrow unittest
 {
     import std.experimental.ndslice.slice;
-    import std.range: iota, retro;
-    auto a = 20.iota.sliced(4, 5);
+    import std.experimental.ndslice.selection: iotaSlice;
+    auto a = iotaSlice(4, 5);
 
     assert(a.dropExactly    !(1, 0)(2, 3)[0, 0] == 17);
     assert(a.dropExactly    !(1, 0)(2, 3).shape == cast(size_t[2])[1, 3]);
```
Comment (wilzbach) [inline std/experimental/ndslice/iteration.d]: why is there so much whitespace here?

**Comment by JackStouffer** [comment]
Besides testing, what's the rationale for iotaSlice?

File: std/experimental/ndslice/iteration.d (hunk 2 above)
Comment (9il) [inline std/experimental/ndslice/iteration.d]: allgned with lines below

**Comment by wilzbach** [comment]
> @wilzbach what do you think about to rename createSlice to slice? We already have for example array and makeArray, ndarray and makeNdarray.

I was actually thinking about this too. Initially I wanted to suggest that we use the same method name for both the gc and allocation version, but prefixing it with `make` does make sense. So +1 for `slice` and a common naming convention.

**Comment by 9il** [comment]
> Besides testing, what's the rationale for iotaSlice?

This is good building block for abstract algorithms along with `indexSlice`.

File: std/experimental/ndslice/slice.d [hunk 3]
```diff
@@ -581,6 +492,319 @@ pure nothrow unittest
...
+    auto slice = array.shape.createSlice!int;
+    slice[] = [[1, 2, 3], [4, 5, 6]];
+    assert(slice == array);
+}
+
+@safe pure unittest
+{
+    size_t[2] shape = (int[][]).init.shape;
+    assert(shape[0] == 0);
+    assert(shape[1] == 0);
```
Comment (wilzbach) [inline std/experimental/ndslice/slice.d]: could you add a test for a higher-dimensional array?
We only test 2d in all tests.

Comment (9il) [inline std/experimental/ndslice/slice.d]: 2D is OK for recursion algorithms. btw, amount of unittests in ndslice already largest in Phobos except `std.datetime`.

Comment (wilzbach) [inline std/experimental/ndslice/slice.d]: > 2D is OK for recursion algorithms

I was just worried that we create a bias with the tests by only testing 2d.

> amount of unittests in ndslice already largest in Phobos except std.datetime

How did you measure this?

Last time I checked there were [some parts](http://forum.dlang.org/thread/vzmkecogfcpoojauqpxw@forum.dlang.org) in Phobos with 100% coverage.

Comment (9il) [inline std/experimental/ndslice/slice.d]: > Last time I checked there were some parts in Phobos with 100% coverage.

I mean amount of LOC

Comment (9il) [inline std/experimental/ndslice/slice.d]: > How did you measure this?

I just know probos pretty well)

**Comment by wilzbach** [comment]
needs rebase :(

**Comment by ZombineDev** [comment]
@9il I have an alternative idea about the allocation API: 
Merge `ndarray` and `makeNdarray` into a single function and use `alias` template parameter for the allocator:
```
// Before:
auto ndarray(size_t N, Range)
    (auto ref Slice!(N, Range) slice);
auto makeNdarray(T, Allocator, size_t N, Range)
    (auto ref Allocator alloc,  Slice!(N, Range) slice);

// After:
auto makeNdarray(alias allocator = GCAllocator, size_t N, Range)
    (auto ref Slice!(N, Range) slice)
{
    alias alloc = allocatorInstanceOf!(allocator);
    alias T = ElementType!Range;

    // same as before...
}

auto m1 = slice.makeNdarray;
auto m2 = slice.makeNdarray!MmapAllocator;
auto m3 = slice.makeNdarray!theAllocator;
auto m4 = slice.makeNdarray!(Mallocator.instance);
```

(Ditto for `makeSlice`).

Full example: http://dpaste.dzfl.pl/c3894699a269

**Comment by wilzbach** [comment]
> Merge ndarray and makeNdarray into a single function and use alias template parameter for the allocator:

We tried this before, but unfortunately `GCAllocator` isn't pure :/

**Comment by 9il** [comment]
> Merge ndarray and makeNdarray into a single function and use alias template parameter for the allocator:
> We tried this before, but unfortunately GCAllocator isn't pure :/

In addition, current API is similar to `array` and `makeArray`+`dispose`. `ndarray` is good name, btw)

**Comment by ZombineDev** [comment]
@wilzbach
> We tried this before, but unfortunately GCAllocator isn't pure :/

Then we should fix GCAllocator ;) I find `std.experimental.allocator` surprisingly lacking in terms of attributes. I did [add](https://github.com/D-Programming-Language/phobos/pull/3957) some, but I didn't look at `gc_allocator` and `mmap_allocator` because I wasn't needing them at the time.  `pure` is going to be the hard one because `Allocator.instance` is mutable and shared. On the other hand, it is debatable whether this attribute should be used for anything but `GCAllocator.allocate`.

@9il
> In addition, current API is similar to array and makeArray+dispose. ndarray is good name, btw)

OK, but what do you think about only changing `makeNdarray` to the the API I proposed, which IMO is easier / more flexible to use:
```
auto makeNdarray(alias allocator = GCAllocator, size_t N, Range)
    (auto ref Slice!(N, Range) slice);

auto m1 = slice.makeNdarray;
auto m2 = slice.makeNdarray!MmapAllocator;
auto m3 = slice.makeNdarray!theAllocator;
auto m4 = slice.makeNdarray!(Mallocator.instance);
```

**Comment by 9il** [comment]
> OK, but what do you think about only changing makeNdarray to the the API I proposed, which IMO is easier / more flexible to use:

If @andralex change it for `makeArray` too.

**Comment by wilzbach** [comment]
> Then we should fix GCAllocator ;) [...] pure is going to be the hard one because Allocator.instance is mutable and shared. On the other hand, it is debatable whether this attribute should be used for anything but GCAllocator.allocate.

I submitted #4190 which hopefully explains the problem that requires us to provide two functions for everything that uses the allocation API.

**Comment by 9il** [comment]
@DmitryOlshansky Could you please merge this. It is blocker for other fixes ( #4203 ) and Mir.

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by 9il** [comment]
Thanks!

**Pull request merged**